File: Solutions/Python/merge_sorted_array.py
"""
    [EASY]
    88. Merge Sorted Array

    Concepts:
    - two pointers
    - sorting

    Stats:
        Runtime | 0 ms      [Beats 100%]
        Memory  | 16.74 MB  [Beats 87.30%]
"""

class Solution:
    def merge(self, nums1: List[int], m: int, nums2: List[int], n: int) -> None:
        """
        Do not return anything, modify nums1 in-place instead.
        """
        ptr = 0
        while len(nums1) > m and nums1[-1] == 0:
            nums1.pop(-1)

        while nums2:
            val = nums2[0]
            while ptr < len(nums1) and nums1[ptr] < val:
                ptr += 1
            nums1.insert(ptr, val)
            nums2 = nums2[1:]

File: Solutions/Cpp/merge_sorted_array.cpp
/*
    [EASY]
    88. Merge Sorted Array

    Concepts:
    - array
    - two pointers

    Stats:
        Runtime | 0 ms      [Beats 100%]
        Memory  | 12.20 MB  [Beats 70.67%]
*/

class Solution {
public:
    void merge(vector<int>& nums1, int m, vector<int>& nums2, int n) {
        int i = 0, j = 0;
        while (i < m + n && j < n) {
            if (nums1[i] > nums2[j]) {
                nums1.insert(nums1.begin() + i, nums2[j]);
                j++;
            }
            else {
                i++;
            }
        }
        while (j < n) {
            nums1.insert(nums1.begin() + m + j, nums2[j]);
            j++;
        }
        nums1.erase(nums1.begin() + m + n, nums1.end());
    }
};